File: tests/Rask.DevTools.E2E.Tests/Apps/Islands/DevToolsMeter.ts
// A plain custom element, used as a Lit-runtime island: no npm import, no decorators (Vite's transform does not lower
// them, and an element that never upgrades renders empty with nothing said).
//
// The adapter assigns each prop as a property while the island mounts, inside the try that reports a mount failure to
// the devtools. So `broken = true` throwing from its setter is a real island mount failure, not a page error.
import type { DevToolsMeterProps } from '@rask/DevToolsMeter.props'

class DevToolsMeterElement extends HTMLElement {
  #label: DevToolsMeterProps['label'] = ''
  #value: DevToolsMeterProps['value'] = 0

  set label(value: string) {
    this.#label = value
    this.#render()
  }

  set value(value: number) {
    this.#value = value
    this.#render()
  }

  set broken(value: boolean) {
    if (value) {
      throw new Error('the meter island failed on purpose')
    }
  }

  connectedCallback(): void {
    this.#render()
  }

  #render(): void {
    this.textContent = `${this.#label}: ${this.#value}`
  }
}

customElements.define('devtools-e2e-meter', DevToolsMeterElement)

export default 'devtools-e2e-meter'

File: tests/Rask.DevTools.E2E.Tests/Apps/Islands/DevToolsMeter.cs
namespace Rask.DevTools.E2E.Tests.Apps.Islands;

/// <summary>
///     A Lit-runtime island — a plain custom element, bundled by the build — that the devtools must show as a component
///     row badged <c>Lit</c> with these props, and whose failure they must list.
/// </summary>
public sealed partial class DevToolsMeter : Rask.External.LitComponent
{
    /// <summary>The caption.</summary>
    public required string Label { get; set; }

    /// <summary>The reading.</summary>
    public int Value { get; set; } = 0;

    /// <summary>When true, the element's setter throws as the island mounts: the failure the devtools list.</summary>
    public bool Broken { get; set; } = false;
}
